### PR TITLE
Add blas64 blas wrapper

### PR DESCRIPTION
**Pull request by kortschak**
This is the first of a set of changes to rationalise and centralise BLAS implementation registration.

No tests are added since this is a thin wrapper around the blas.Float64 interface and is tested elsewhere.

### CODE REVIEW

**Comment by kortschak** [comment]
A few things remain open in the design:

1. I am considering changing the names of `goblas.Blas` and `cblas.Blas` to `Implementation`. This has the nice effect that registering the implementation reads as follows: e.g. `blas64.Use(goblas.Implementation{})`. However `Implementation` is quite long. Views?
2. The name of `RawMatrix` and `RawMatrixer`. Should they stay the same?
3. `blas64.General` is named from the BLAS nomenclature which differs from ours. Keep it as is?

I should explain why I did not just use `gonum/blas/dbw`. There is a significant amount of range checking in that package that doubles up on the checks in the blas implementations. Further, there is some flexibility in `blas64` that is not possible in `dbw` that I found was necessary when using it as a client in `mat64`. If no people are using `dbw` it should probably be deleted.

File: blas64/blas64.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+
+// Use sets the BLAS float64 implementation to be used by subsequent BLAS calls.
+// The default implementation is goblas.Blas.
+func Use(b blas.Float64) {
+	blas64 = b
+}
+
+// Implementation returns the current BLAS float64 implementation.
+//
+// Implementation allows direct calls to current the BLAS float64 implementation
```
Comment (vladimir-ch) [inline blas64/blas64.go]: *the* current?

File: blas64/blas64.go [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+}
+
+// General represents a matrix using the conventional storage scheme.
+type General struct {
+	Rows, Cols int
+	Stride     int
+	Data       []float64
+}
+
+// Band represent a band matrix using the band storage scheme.
```
Comment (vladimir-ch) [inline blas64/blas64.go]: represent**s**

File: blas64/blas64.go [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+// Triangular represents a triangular matrix using the conventional storage scheme.
+type Triangular struct {
+	N      int
+	Stride int
+	Data   []float64
+	Uplo   blas.Uplo
+	Diag   blas.Diag
+}
+
+// Triangular represents a triangular matrix using the band storage scheme.
```
Comment (vladimir-ch) [inline blas64/blas64.go]: Triangular**Band**

File: blas64/blas64.go [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+type TriangularBand struct {
+	N, K   int
+	Stride int
+	Data   []float64
+	Uplo   blas.Uplo
+	Diag   blas.Diag
+}
+
+// Triangular represents a triangular matrix using the packed storage scheme.
```
Comment (vladimir-ch) [inline blas64/blas64.go]: Triangular**Packed**

File: blas64/blas64.go [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+
+// Symmetric represents a symmetric matrix using the conventional storage scheme.
+type Symmetric struct {
+	N      int
+	Stride int
+	Data   []float64
+	Uplo   blas.Uplo
+}
+
+// Symmetric represents a symmetric matrix using the band storage scheme.
```
Comment (vladimir-ch) [inline blas64/blas64.go]: SymmetricBand and also below SymmetricPacked

File: blas64/blas64.go [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+}
+
+// Level 1
+
+const (
+	negIncX = "blas64: negative x increment"
+	negIncY = "blas64: negative y increment"
+)
+
+// Dot will panic if either vector increment is negative.
```
Comment (vladimir-ch) [inline blas64/blas64.go]: If *either*, then y.Inc should be checked too.

File: blas64/blas64.go [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+
+func Rotg(a, b float64) (c, s, r, z float64) {
+	return blas64.Drotg(a, b)
+}
+
+func Rotmg(d1, d2, b1, b2 float64) (p blas.DrotmParams, rd1, rd2, rb1 float64) {
+	return blas64.Drotmg(d1, d2, b1, b2)
+}
+
+// X will panic if either vector increment is negative.
```
Comment (vladimir-ch) [inline blas64/blas64.go]: Rot will panic

**Comment by vladimir-ch** [comment]
1) Implementation is quite long, but it won't be written frequently and blas64.Use(goblas.Implementation{}) read so nicely. I would change it.

File: blas64/blas64.go (hunk 6 above)
Comment (kortschak) [inline blas64/blas64.go]: Added.

**Comment by kortschak** [comment]
@vladimir-ch I just tried `git commit --amend` and `git push -f` and we get what we see here. I think this works quite well - we have the review history and the result will be a single commit. I'll wait on others' opinions wrt `Implementation`, I like it too, but it will break code elsewhere.

**Comment by jonlawlor** [comment]
I think `Implementation` is perfect.  If you google search for blas implementation then in the first 2 pages you get the reference, ATLAS, GotoBLAS, OpenBLAS, cuBLAS, and uh, gonum/blas.  I guess you beat Intel and Apple?

**Comment by fhs** [comment]
`goblas.Implementation{}`  is fine, but there's still some stutter in `blas64.Use(goblas.Implementation{})`. I'd prefer `blas64.Use(native.Implementation{})`. I don't know if renaming the package is an option here but you are already breaking compatibility.

**Comment by kortschak** [comment]
As an aid for integrating the design, this is how I see the client's view of the system after the change has gone in.

mat64/blas/blas64 dependency graph

(Dashed lines are optional for the client).

**Comment by kortschak** [comment]
>I'd prefer `blas64.Use(native.Implementation{})`.

That would imply `blas64.Use(c.Implementation{})` (`C` is not available because of th CGO).

**Comment by fhs** [comment]
or `blas64.Use(cgo.Implementation{})`, which might make more sense because the underlying routines might be written in assembly, not just C. You're just using cgo to wrap them.

**Comment by kortschak** [comment]
That's fair. I'm waiting on @btracey's comments since this (and with `Implementation`, `native` and `cgo` would further) touches his work significantly.

File: blas64/blas64.go [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+}
+
+// General represents a matrix using the conventional storage scheme.
+type General struct {
+	Rows, Cols int
+	Stride     int
+	Data       []float64
+}
+
+// Band represents a band matrix using the band storage scheme.
```
Comment (btracey) [inline blas64/blas64.go]: A banded matrix

File: blas64/blas64.go [hunk 9]
```diff
@@ -0,0 +1,331 @@
...
+// Triangular represents a triangular matrix using the conventional storage scheme.
+type Triangular struct {
+	N      int
+	Stride int
+	Data   []float64
+	Uplo   blas.Uplo
+	Diag   blas.Diag
+}
+
+// TriangularBand represents a triangular matrix using the band storage scheme.
```
Comment (btracey) [inline blas64/blas64.go]: Represents a banded triangular matrix using

**Comment by btracey** [comment]
Maybe it should be "general storage scheme"rather than conventional

**Comment by btracey** [comment]
Implementation is great.  I like cgo.Implementation, but I'm not a huge fan of native. I don't have better suggestions at the moment.

File: blas64/blas64.go [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+
+// Iamax will panic if the vector increment is negative.
+func Iamax(n int, x Vector) int {
+	if x.Inc < 0 {
+		panic(negIncX)
+	}
+	return blas64.Idamax(n, x.Data, x.Inc)
+}
+
+// Swap will panic if either vector increment is negative.
```
Comment (btracey) [inline blas64/blas64.go]: Swap, copy, axpy, etc panics are incorrect. With negative vectors they do what you'd expect,  and I assume they are useful given they are in blas. It's the single vector cases that have surprising behavior with a negative increment, and this should panic.

File: blas64/blas64.go (hunk 8 above)
Comment (kortschak) [inline blas64/blas64.go]: I wasn't sure of this so I did a survey and it seems to be pretty much equally either way; band is shorter, and netlib and wikipedia both call it "band matrix" (with wikipedia saying "or banded".

File: blas64/blas64.go (hunk 9 above)
Comment (kortschak) [inline blas64/blas64.go]: As above.

File: blas64/blas64.go (hunk 10 above)
Comment (kortschak) [inline blas64/blas64.go]: OK, I'll fix that. Thanks.

**Comment by kortschak** [comment]
>Maybe it should be "general storage scheme"rather than conventional

This is the netlib terminology. Add a link?

>Implementation is great. I like cgo.Implementation, but I'm not a huge fan of native. I don't have better suggestions at the moment. 

I'm happy to wait. I'll do both changes at the same time, but I'd rather do them all only once.

File: blas64/blas64.go (hunk 8 above)
Comment (btracey) [inline blas64/blas64.go]: Okay.

**Comment by btracey** [comment]
LGTM. 

I don't like native because people talk about having a "native" implementation, which usually implies assembly and all of the tricks. Not that that's bad, but ours is and likely will continue to be a mix of go and assembly. How about "standard"?

````
blas64.Register(standard.Implementation{})
````

reads pretty well.

**Comment by fhs** [comment]
I have no strong attachment to "native". "standard" is fine or even "std". I just don't like the stutter.

Just curious, why is this package called "blas64"? Shouldn't it be "float64" or "f64"? Go also has a complex64 type.

**Comment by kortschak** [comment]
It can't be `float64` because of the builtin type, but it could be `f64`or `real64`. My original intention was to name the complex package `cmplxblas128`, but `cmplx128` would be OK if one of the others were chosen for the float64 package.

I don't think `native` has the implementation of assembly (for example [here](http://golang.org/doc/faq#What_compiler_technology_is_used_to_build_the_compilers): "That plan aside, Go is a fine language in which to implement a self-hosting compiler: a *native* lexer and parser are already available in the go package and a separate type checking package has also been written. "). I actually prefer it to `standard` or `std`, but of those two I prefer `std` because of length.

**Comment by jonlawlor** [comment]
You can't call it `go`, but you could call it `golang`.

**Comment by btracey** [comment]
Native is fine then. 

blas64 is the name at the moment because of the parallel with mat64. mat64 is named as such because as far as I'm aware we don't intend to ever have a matrix package for float32 or cmplx64. Blas is potentially different because the functions are there, though if in general gonum will only have float64 and cmplx128 types, then there's also no need for convenience wrappers for the other numeric types. In general, I'm not a fan of f64 because it's a relatively uninformative short package name (and it's much easier in go if you don't have to give renames), but in the case of BLAS it's probably okay because the hope is that people won't have to interact with it much (and instead we'll just make matrix better).

**Comment by kortschak** [comment]
PTAL

There are residual references to goblas in panic messages, but these will be cleaned up in part 2 of the panic harmonisation.

I would like to minimise the window of skew between mat64 and blas64 caused by this, so I won't merge this until gonum/matrix#66 has been reviewed. (Tests pass on my machine FWIW).

**Comment by btracey** [comment]
LGTM

**Pull request merged**